### ISSUE TITLE
Old VLAN tag remains on VF when NAD has no VLAN field.

### ISSUE THREAD

**Issue by alonSadan** [openshift/sriov-cni#25]
### What happened?
Once a SRIOV VF gets a VLAN tag assigned to,the vlan tag vlaue is kept,
even though the vmi/pod was connected to a network with no VLAN attribute defined on it's `NetworkAttachmentDefenition`.

### What did you expect to happen?
When VLAN not specified - the vlan tag will be removed from the VF.
When  VLAN equals 0 - the vlan tag should be with value 0.
### What are the minimal steps needed to reproduce the bug?
On an SRIOV setup: 

1) set a VF with a vlan tag
2) create a NAD with no vlan attribute for example: 

```
apiVersion: k8s.cni.cncf.io/v1
kind: NetworkAttachmentDefinition
metadata:
  name: "sriov-no-vlan"
  namespace: "default"
  annotations:
    k8s.v1.cni.cncf.io/resourceName: "openshift.io/sriov_net"
spec:
  config: >-
    { "cniVersion": "0.3.1", "name": "sriov-network", "type": "sriov", "link_state":
    "enable", "ipam": {}}
```
3) create a pod (could be vmi) and connect it to the VLANless network. I used: 
```
apiVersion: v1
kind: Pod
metadata:
  name: example-pod-no-vlan-4
  annotations:
    k8s.v1.cni.cncf.io/networks: sriov-no-vlan 
spec:
  containers:
  - name: example-pod
    command: ["/bin/bash", "-c", "sleep 2000000000000"]
    image: centos/tools
```
4) make sure the VF that gets assigned is the one that got a vlan, I just gave all VFs the same random VLAN tag.

### Anything else we need to know?
When The vlan attribute exits on the NAD, but holds a value of `0` the VLAN does get removed from the VF.

### Component Versions
Please fill in the below table with the version numbers of applicable components used.

Component | Version|
------------------------------|--------------------|
|SR-IOV CNI Plugin            |4.4|
|Multus                       |<Input Version Here>|
|SR-IOV Network Device Plugin |4.4|
|Kubernetes                   |kind-k8s-1.17| 
|OS                           |CentOS Linux 7|

### Config Files
Config file locations may be config dependent.

**Comment by zshi-redhat** [comment]
@alonSadan so you would like to see the VF vlan tag be set to 0 when not explicitly specified in net-attach-def?

**Comment by alonSadan** [comment]
As @EdDev said [here](https://github.com/kubevirt/kubevirt/pull/5093#issuecomment-785202879), and @oshoval said [here](https://github.com/kubevirt/kubevirt/pull/5093#pullrequestreview-598377245), there is a difference between vlan=0 and no vlan at all. We would like it to be reflected when applying the NAD:
- When vlan is net to 0 on the NAD , the VF will have vlan=0. 
- When vlan is not specified in the NAD , the VF will have no vlan tag at all. 

Is this possible?

**Comment by zshi-redhat** [comment]
@adrianchiris @martinkennelly Is there any difference between setting vlan=0 and not setting any vlan on VFs?

**Comment by zshi-redhat** [comment]
> As @EdDev said [here](https://github.com/kubevirt/kubevirt/pull/5093#issuecomment-785202879), and @oshoval said [here](https://github.com/kubevirt/kubevirt/pull/5093#pullrequestreview-598377245), there is a difference between vlan=0 and no vlan at all. We would like it to be reflected when applying the NAD:
> 
>     * When vlan is net to 0 on the NAD , the VF will have vlan=0.
> 
>     * When vlan is not specified in the NAD , the VF will have no vlan tag at all.

I think these two are already satisfied in sriov-cni, isn't it?
However, I noticed that in SR-IOV Operator, it adds default vlan=0 in net-attach-def if not explicitly specified.

**Comment by EdDev** [comment]
> > ```
> > * When vlan is net to 0 on the NAD , the VF will have vlan=0.
> > 
> > * When vlan is not specified in the NAD , the VF will have no vlan tag at all.
> > ```
> 
> I think these two are already satisfied in sriov-cni, isn't it?

Per my understanding of @alonSadan findings, if the VLAN is not specified, the VF is not touching that property, therefore leaving what was there before. This is a different semantic from what was mentioned above.

> However, I noticed that in SR-IOV Operator, it adds default vlan=0 in net-attach-def if not explicitly specified.

**Comment by alonSadan** [comment]
> > As @EdDev said [here](https://github.com/kubevirt/kubevirt/pull/5093#issuecomment-785202879), and @oshoval said [here](https://github.com/kubevirt/kubevirt/pull/5093#pullrequestreview-598377245), there is a difference between vlan=0 and no vlan at all. We would like it to be reflected when applying the NAD:
> > ```
> > * When vlan is net to 0 on the NAD , the VF will have vlan=0.
> > 
> > * When vlan is not specified in the NAD , the VF will have no vlan tag at all.
> > ```
> 
> I think these two are already satisfied in sriov-cni, isn't it?
> However, I noticed that in SR-IOV Operator, it adds default vlan=0 in net-attach-def if not explicitly specified.

From what I saw, When VLAN is not specified in the NAD, the VLAN attribute of the VF will not change, but we expect it to be removed.

**Comment by martinkennelly** [comment]
> @adrianchiris @martinkennelly Is there any difference between setting vlan=0 and not setting any vlan on VFs?

@zshi-redhat No functional difference for X710 and E810s. 
@alonSadan What driver and hardware are you using?

**Comment by adrianchiris** [comment]
>> @adrianchiris @martinkennelly Is there any difference between setting vlan=0 and not setting any vlan on VFs?

**Comment by zshi-redhat** [comment]
> > > As @EdDev said [here](https://github.com/kubevirt/kubevirt/pull/5093#issuecomment-785202879), and @oshoval said [here](https://github.com/kubevirt/kubevirt/pull/5093#pullrequestreview-598377245), there is a difference between vlan=0 and no vlan at all. We would like it to be reflected when applying the NAD:
> > > ```
> > > * When vlan is net to 0 on the NAD , the VF will have vlan=0.
> > > 
> > > * When vlan is not specified in the NAD , the VF will have no vlan tag at all.
> > > ```
> > 
> > 
> > I think these two are already satisfied in sriov-cni, isn't it?
> > However, I noticed that in SR-IOV Operator, it adds default vlan=0 in net-attach-def if not explicitly specified.
> 
> From what I saw, When VLAN is not specified in the NAD, the VLAN attribute of the VF will not change, 

Ok, so you are expecting sriov-cni to set vlan=0 on VF when vlan field is not specified in NAD.
When sriov-cni is used in SR-IOV Operator, NAD is created with vlan=0 by default via SR-IOV Operator API.
When sriov-cni is not used w/ SR-IOV Operator, it by default will ignore the vlan field if not configured, meaning vlan=0 is not set on VF device.

Given that Martin and Adrian had confirmed that `vlan=0` is equal to `no vlan`, I think it doesn't hurt if we change sriov-cni to always set `vlan=0` when vlan is not configured in NAD.

**Comment by adrianchiris** [comment]
an "unspecified" VLAN in NAD does not necessarily mean vlan 0. I expect it to mean unspecified as in the CNI plugin should not concern itself with VLAN configuration.

**Comment by EdDev** [comment]
> an "unspecified" VLAN in NAD IMO does not necessarily mean vlan 0. I expect it to mean "unspecified" as in the CNI plugin should not concern itself with VLAN configuration.

A VLAN or lack of it defines a network, it is unreasonable to get a random one. When one does not specify a VLAN, he explicitly says that it expects it to have none.
Saying something like: Please assign me a VF, no matter what VLAN it is, makes no sense.

**Comment by adrianchiris** [comment]
> A VLAN or lack of it defines a network, it is unreasonable to get a random one.

It is the VLAN as was configured in the system it is not "random".

**Comment by EdDev** [comment]
> > A VLAN or lack of it defines a network, it is unreasonable to get a random one.
> 
> It is the VLAN as was configured in the system it is not "random".

That is not reasonable IMO, the CNI assigns and releases VF/s and basically controls its settings.
If the NAD is not specifying a VLAN, I conclude, as with classical networking, that there is non. Asking the operator not to trust that because the NAD is "merged" with how the VF/s have been setup by the HW admin is unreasonable.
There is a meaning of no-VLAN and such a semantic behavior is both confusing and contradicts other implementations (e.g. the SR-IOV Operator).

At KubeVirt we have been experiencing flakiness in tests due to this. The VF/s assigned had on them leaked VLAN/s (probably from other tests that used them, and somehow the release has not restored VLAN 0), and at some point such VF/s have been assigned to pods who expected no VLAN.

**Comment by alonSadan** [comment]
@EdDev  @adrianchiris @zshi-redhat 

What I understand from the discussion is that two things needs to be done in order to close this issue: 

1) update the code so that when no VLAN is provided, the VLAN will be removed (treat it like VLAN ID = 0)
2) give a note about (1) in the docs.

Any objections?

**Comment by alonSadan** [comment]
Hi @zshi-redhat @adrianchiris  @martinkennelly @EdDev 
I added a [PR](https://github.com/openshift/sriov-cni/pull/29) to: 

> 1) update the code so that when no VLAN is provided, the VLAN will be removed (treat it like VLAN ID = 0)
>   2) give a note about (1) in the docs.

**Comment by alonSadan** [comment]
[U/S PR](https://github.com/k8snetworkplumbingwg/sriov-cni/pull/180) got merged , and the 4.9 D/S release already include the solution for this issue.
I think this one can be closed.
